src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+
+            if (unconvertedInterpolatedString.ConstantValue is not null)
+            {
+                // Case 1
+                Debug.Assert(unconvertedInterpolatedString.Parts.All(static part => part.Type is null or { SpecialType: SpecialType.System_String }));
+                return constructWithData(BindInterpolatedStringParts(unconvertedInterpolatedString, diagnostics), data: null);
+            }
+
+            if (tryBindAsBuilderType(out var result))
+            {
+                // Case 2
+                return result;
+            }
+
+            // The specifics of 3 vs 4 aren't necessary for this stage of binding. The only thing that matters is that every part needs to be convertible
+            // object.
+            return constructWithData(BindInterpolatedStringParts(unconvertedInterpolatedString, diagnostics), data: null);
+
+            BoundInterpolatedString constructWithData(ImmutableArray<BoundExpression> parts, InterpolatedStringBuilderData? data)
+                => new BoundInterpolatedString(
+                    unconvertedInterpolatedString.Syntax,
+                    data,
+                    parts,
+                    unconvertedInterpolatedString.ConstantValue,
+                    unconvertedInterpolatedString.Type,
+                    unconvertedInterpolatedString.HasErrors);
+
+            bool tryBindAsBuilderType([NotNullWhen(true)] out BoundInterpolatedString? result)
+            {
+                if (this.Flags.Includes(BinderFlags.InCatchFilter))
+                {
+                    // CLI Spec does not permit nested tries inside a filter block, so we can't use the builder here, as it is disposable.
+                    // PROTOTYPE(interp-string): Should we try to collect errors if a type not otherwise usable in an interpolated string
+                    // is used in this case?
+                    result = null;
+                    return false;
+                }
+
+                result = null;
+                var interpolatedStringBuilderType = Compilation.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_InterpolatedStringBuilder);
+                if (interpolatedStringBuilderType is MissingMetadataTypeSymbol)
+                {
+                    return false;
+                }
+
+                if (unconvertedInterpolatedString.Parts.ContainsAwaitExpression())
+                {
+                    return false;
+                }
+
+                BindingDiagnosticBag applicableDiagnostics = BindingDiagnosticBag.GetInstance(template: diagnostics);
+                if (!Conversions.IsApplicableInterpolatedStringBuilderType(unconvertedInterpolatedString, interpolatedStringBuilderType, applicableDiagnostics, out var builderArguments))
+                {
+                    applicableDiagnostics.Free();
+                    return false;
```

Comment:
>return false; [](start = 20, length = 13)

I am not sure whether this is a design question or more an implementation question, but I would prefer the implementation and the rules be simpler. In particular, If the builder type is there, and other conditions not related to the implementation of the builder type are met (lack of awaits, not in catch filter, etc.), I think we should bind interpolation as a builder. Any unexpected situations we run into (lack of overloads or inconsistent overloads, etc.) should be treated as errors and be reported as such. Otherwise, the implementation feels fragile, it is not clear what errors should be treated as real errors and what errors should be ignored along with the type. For consumer, it is hard to figure out why the builder type is getting ignored when it looks like it has everything it needs.

Reply:
This will partially depend on the builder definition we decide on in the framework. With the current plan, we should able to support any expression type aside from things that will never be supported no matter whether we do `string.Format` or the builder, like pointer types. If we confirm that design decision tomorrow, I think we can simplify this code.

---
In reply to: [610810182](https://github.com/dotnet/roslyn/pull/52422#discussion_r610810182) [](ancestors = 610810182)